File: ResourceManager/ViewModels/Notes/NotePopupViewModel.cs
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using NetCore.Mvvm.ViewModels;
using NetCore.Security;
using NetCore.Validators;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using System;
using System.Threading.Tasks;
using System.Windows.Input;

namespace ResourceManager.ViewModels
{
    public class NotePopupViewModel : PopupViewModel
    {
        private IWindowManager windowManager;
        private readonly IDataServiceManager _dataServiceManager;
        private int ID;
        private bool isEdit;

        public NotePopupViewModel(IDataServiceManager dataServiceManager, IWindowManager windowManager)
        {
            _dataServiceManager = dataServiceManager;
            this.windowManager = windowManager;
            isEdit = false;

            AddValidationRule(() => Title, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Title)));
            AddValidationRule(() => Notes, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Notes)));
            
            Date = DateTime.Now.Date;

            SaveCommand = new AsyncRelayCommand(OnSaveAsync, CanSave);
        }

        private bool CanSave()
        {
            return !HasErrors;
        }
        public void SetNote(Note note)
        {
            isEdit = true;
            ID = note.ID;
            Title = note.Title;
            Date = note.Date;
            Notes = note.Notes;
        }
        private async Task OnSaveAsync()
        {
            ValidateAllRules();

            if (!HasErrors)
            {
                try
                {
                    var note = new Note()
                    {
                        Title = Title,
                        Date = Date.Date,
                        Notes = Notes,
                        UserID = AuthManager.User.Identity.ID
                    };

                    if (isEdit)
                    {
                        await _dataServiceManager.NoteDataService.Update(ID, note);

                        Result = PopupResult.OK;
                    }
                    else
                    {
                        await _dataServiceManager.NoteDataService.Create(note);

                        Result = PopupResult.OK;
                    }
                }
                catch (Exception e)
                {
                    windowManager.ShowMessageBox(e.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
                }
                OnClose();
            }
        }

        public string Title
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public DateTime Date
        {
            get => GetProperty<DateTime>();
            set => SetProperty(value);
        }
    
        public string Notes
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public ICommand SaveCommand { get; }
    }
}

File: ResourceManager/ViewModels/Notes/NotePopupViewModel.vb
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports VBNetCore.Mvvm.ViewModels
Imports VBNetCore.Security
Imports VBNetCore.Validators
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions
Imports System
Imports System.Threading.Tasks
Imports System.Windows.Input

Namespace ResourceManager.ViewModels
    Public Class NotePopupViewModel
        Inherits PopupViewModel
        Private windowManager As IWindowManager
        Private ReadOnly _dataServiceManager As IDataServiceManager
        Private ID As Integer
        Private isEdit As Boolean

        Public Sub New(ByVal dataServiceManager As IDataServiceManager, ByVal windowManager As IWindowManager)
            _dataServiceManager = dataServiceManager
            Me.windowManager = windowManager
            isEdit = False

            AddValidationRule(Function() Title, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(NotePopupViewModel.Title)))
            AddValidationRule(Function() Notes, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(NotePopupViewModel.Notes)))

            [Date] = Date.Now.Date

            SaveCommand = New AsyncRelayCommand(AddressOf OnSaveAsync, AddressOf CanSave)
        End Sub

        Private Function CanSave() As Boolean
            Return Not HasErrors
        End Function
        Public Sub SetNote(ByVal note As Note)
            isEdit = True
            ID = note.ID
            Title = note.Title
            [Date] = note.Date
            Notes = note.Notes
        End Sub
        Private Async Function OnSaveAsync() As Task
            ValidateAllRules()

            If Not HasErrors Then
                Try
                    Dim note = New Note() With {
    .Title = Title,
    .[Date] = [Date].Date,
    .Notes = Notes,
    .UserID = AuthManager.User.Identity.ID
}

                    If isEdit Then
                        Await _dataServiceManager.NoteDataService.Update(ID, note)

                        Result = PopupResult.OK
                    Else
                        Await _dataServiceManager.NoteDataService.Create(note)

                        Result = PopupResult.OK
                    End If
                Catch e As Exception
                    windowManager.ShowMessageBox(e.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error)
                End Try
                OnClose()
            End If
        End Function

        Public Property Title As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property [Date] As Date
            Get
                Return GetProperty(Of Date)()
            End Get
            Set(ByVal value As Date)
                SetProperty(value)
            End Set
        End Property

        Public Property Notes As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public ReadOnly Property SaveCommand As ICommand
    End Class
End Namespace
